### DISCUSSION TITLE
HTML support in Tooltip

### DISCUSSION THREAD

**Discussion by thetableman** [zauberzeug/nicegui#1596] [Ideas / Feature Requests]
The current `ui.tooltip` (based on a `TextElement` class) doesn't allow for HTML or the use of `\n` for new lines.

Is there any reason the element can't be based on the `ContentElement` class which would support this?

```py
from nicegui import ui

with ui.button('Example'):
    ui.tooltip('Test <i>Tooltip!</i><br>Now with <b>HTML</b> and multi-line support!').classes('text-body2')
    
ui.run()
```
Current output using `TextElement`
![image](https://github.com/zauberzeug/nicegui/assets/78121198/d280c887-99e5-4aed-abd8-27009bc4d4e0)

Potential output using `ContentElement`
![image](https://github.com/zauberzeug/nicegui/assets/78121198/e0d43c3c-4815-4d78-ab6c-80b914ca9919)

**Comment by rodja** [comment]
Interesting observation @thetableman. I would really like to see this. But changing the inheritance will mean a breaking change which we should only do in 1.4. Another alternative may be to also use `innerHTML` in the TextElement. In 1.4 we could then remove the duplicity and go with either TextElement or ContentElement for all Elements. I'm not sure why the distinction was made in the first place. @falkoschindler will be back at the end of the week and may shed some light into this.

**Reply by thetableman** [reply]
There could be some breaking changes I haven't identified but after updating the `ui.tooltip` class in an active project as below I haven't experienced any issues. I think the breaking changes would come from changing the name of the `text` parameter to `content` - this is obviously import for consistency in naming across all elements but also not actually necessary for the changes to work with the element.

```py
# from .mixins.text_element import TextElement
from .mixins.content_element import ContentElement


# class Tooltip(TextElement):
class Tooltip(ContentElement):

    def __init__(self, text: str) -> None:
        """Tooltip

        Can be placed in another element to show additional information on hover.

        :param text: the content of the tooltip
        """
        # super().__init__(tag='q-tooltip', text=text)
        super().__init__(tag='q-tooltip', content=text)
```

**Reply by rodja** [reply]
But the helper methods like `set_text` will change to `set_content` and may break existing code in user codebases.

**Comment by thetableman** [comment]
> But the helper methods like `set_text` will change to `set_content` and may break existing code in user codebases.

That is true, I didn't consider the impacts that broadly because they weren't relevant to my current project.

**Comment by falkoschindler** [comment]
There you go:

```py
with ui.button('Example'):
    with ui.tooltip('').classes('text-body2'):
        ui.html('Test <i>Tooltip!</i><br>Now with <b>HTML</b> and multi-line support!')
```

The only imperfection is the need for the tooltip text `''`. We could make this parameter optional.

**Reply by falkoschindler** [reply]
We just released version 1.4.6 with the optional `text` parameter. So I think this feature request is done. 🙂

**Comment by falkoschindler** [comment]
Regarding the distinction between text and content: Text is meant to be displayed as is, without translating `<` as HTML or `*` as Markdown. But for certain elements, we _expect_ the content to be more than text. Then it makes sense to interpret it accordingly.
Supporting HTML in `ui.tooltip()` directly would mean you'd always need to escape characters like `<`, which would be a huge hassle!
